src/Symfony/Component/Uid/ArgumentResolver/UidValueResolver.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Uid\ArgumentResolver;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Controller\ArgumentValueResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use Symfony\Component\Uid\AbstractUid;
+use Symfony\Component\Uid\Attribute\Uid;
+
+final class UidValueResolver implements ArgumentValueResolverInterface
+{
+    private string $defaultFormat;
+
+    public function __construct(string $defaultFormat = Uid::FORMAT_CANONICAL)
+    {
+        $this->defaultFormat = $defaultFormat;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(Request $request, ArgumentMetadata $argument): bool
+    {
+        return !$argument->isVariadic() &&
+            \is_string($request->attributes->get($argument->getName())) &&
+            null !== $argument->getType() &&
+            is_a($argument->getType(), AbstractUid::class, true);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function resolve(Request $request, ArgumentMetadata $argument): iterable
+    {
+        $format = $this->defaultFormat;
+        if (null !== $configuration = method_exists($argument, 'getAttributes') ? $argument->getAttributes(Uid::class, ArgumentMetadata::IS_INSTANCEOF)[0] ?? null : null) {
+            if (null !== $configuration->format) {
+                $format = $configuration->format;
+            }
+        }
+
+        /** @var class-string<AbstractUid> $uidClass */
+        $uidClass = $argument->getType();
+
+        $c = match ($format) {
+            Uid::FORMAT_CANONICAL => static fn (string $uid): AbstractUid => new $uidClass($uid), // TODO: this is wrong
```

Comment:
@nicolas-grekas Est-ce qu'on peut rajouter `abstract public static function fromCanonical(string $uid): static;` dans `AbstractUid`, je ne vois pas comment instancier les classes custom qui extend direct `AbstractUid`  au format canonical (sans le connaître) ? Le constructeur ne faisant pas partie du contrat de `AbstractUid`.

Reply:
Voir https://github.com/fancyweb/symfony/commit/fce7a6173481d0e8eb57247056d19e68ade9c9b7

Reply:
check after the creating that __toString() === $uid?